src/Sylius/Bundle/CoreBundle/EventListener/OrderTotalIntegrityChecker.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\EventListener;
+
+use Doctrine\Common\Persistence\ObjectManager;
+use Sylius\Bundle\ResourceBundle\Event\ResourceControllerEvent;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Order\Processor\OrderProcessorInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\Routing\RouterInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+final class OrderTotalIntegrityChecker
+{
+    /**
+     * @var OrderProcessorInterface
+     */
+    private $orderProcessors;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var ObjectManager
+     */
+    private $manager;
+
+    /**
+     * @param OrderProcessorInterface $orderProcessors
+     * @param RouterInterface $router
+     * @param ObjectManager $manager
+     */
+    public function __construct(
+        OrderProcessorInterface $orderProcessors,
+        RouterInterface $router,
+        ObjectManager $manager
+    ) {
+        $this->orderProcessors = $orderProcessors;
+        $this->router = $router;
+        $this->manager = $manager;
+    }
+
+    /**
+     * @param ResourceControllerEvent $event
+     */
+    public function check(ResourceControllerEvent $event)
+    {
+        /** @var OrderInterface $order */
+        $order = $event->getSubject();
+
+        Assert::isInstanceOf($order, OrderInterface::class);
+
+        $oldTotal = $order->getTotal();
+        $this->orderProcessors->process($order);
```

Comment:
We would need to perform deep clone to ensure we do not modify anything from the real order. For now we can only clone payments, shipments, order items and order item units of every order item. The adjustments are always removed and new ones added, so can be skipped for the first iteration and the rest of the entities is practically read only.

The following snippet should work for `Order`:
```php
public function __clone()
    {
        $paymentsToClone = null === $this->payments ? new ArrayCollection() : $this->payments;
        $this->payments = new ArrayCollection();
        foreach ($paymentsToClone as $payment) {
            $this->payments->add(clone $payment);
        }

        $shipmentsToClone = null === $this->shipments ? new ArrayCollection() : $this->shipments;
        $this->shipments = new ArrayCollection();
        foreach ($shipmentsToClone as $shipment) {
            $this->shipments->add(clone $shipment);
        }

        $itemsToClone = null === $this->items ? new ArrayCollection() : $this->items;
        $this->items = new ArrayCollection();
        foreach ($itemsToClone as $item) {
            $this->items->add(clone $item);
        }
```

Similar thing for `OrderItem`. It's risky that we can miss something and recalculate only part of the real order and that can cause tragic effect, so I would just remove the `flush` from there for now (as mentioned below) and introduce this change as soon as we will have time to implement it properly (probably after 1.0.0-beta.2 tag). Actually it will not be a BC break.